executor/cook/command.py

```diff
@@ -0,0 +1,83 @@
+import os
+import shlex
+import subprocess
+import time
+
+from collections import namedtuple
+
+Command = namedtuple(
+    'Command',
+    ['id', 'value', 'async', 'guard']
+)
+
+RunningCommand = namedtuple(
+    'RunningCommand',
+    Command._fields + ('stdout', 'stderr', 'process', )
+)
+
+def run_command(command, env):
+    stdout = open('stdout' + str(command.id), 'w+')
+    stderr = open('stderr' + str(command.id), 'w+')
+
+    return RunningCommand(
+        *command,
+        stdout = stdout,
+        stderr = stderr,
+        process = subprocess.Popen(
+            shlex.split(command.value),
+            stdout = stdout,
+            stderr = stderr,
+            env = dict(os.environ, **env)
+        )
+    )
+
+def kill_command(command):
+    command.stdout.close()
+    command.stderr.close()
+
+    if command.process.returncode is None:
+        command.process.terminate()
+        for _ in range(0, 10):
+            if command.process.poll() is not None:
+                break
+            time.sleep(1)
+        if command.process.returncode is None:
+            command.process.kill()
+
+    return command.process.returncode
+
+def is_guard(command):
+    return command.guard or False
+
+def is_async(command):
+    return command.async or False
+
+def is_sync(command):
+    return not is_async(command)
+
+def is_running(command):
+    return command.process.poll() is None
+
+def is_crashed(command):
+    return not is_running(command) and command.process.poll() is not 0
+
+def run_commands(commands, stop = None, get_env = lambda: {}):
+    rcommands = []
+    icommands = iter(commands)
+
+    try:
+        while not (stop and stop.isSet()):
+            should_await = [c for c in rcommands if is_running(c) and is_sync(c)]
+            should_abort = [c for c in rcommands if is_crashed(c) and is_guard(c)]
+
+            if should_abort:
+                break
+            elif not should_await:
+                if len(commands) == len(rcommands):
```

Comment:
What am I missing.. 

If the last command is_running and async, then won't we end up exiting while it is still running?